src/base/cub_ddl_log.c

```diff
@@ -0,0 +1,927 @@
+  if (path[0] == '/')
+    {
+      p = path + 1;
+    }
+#endif /* WINDOWS */
+
+  while (p != NULL)
+    {
+      p = strchr (p, '/');
+      if (p == NULL)
+	return 0;
+
+      if (p != NULL)
+	*p = '\0';
+
+      if (access (path, F_OK) < 0)
+	{
+	  if (mkdir (path, 0777) < 0)
+	    {
+	      return -1;
+	    }
+	}
+      if (p != NULL)
+	{
+	  *p = '/';
+	  p++;
+	}
+    }
+  return 0;
+}
+
+static int
+cub_get_current_time (char *buf, size_t size)
+{
+  struct tm at_tm;
+  int len = 0;
+  time_t t = time (NULL);
+
+  localtime_r (&t, &at_tm);
+  at_tm.tm_year += 1900;
+  at_tm.tm_mon += 1;
+  len =
+    snprintf (buf, size, "%04d%02d%02d_%02d%02d%02d", at_tm.tm_year, at_tm.tm_mon, at_tm.tm_mday, at_tm.tm_hour,
+	      at_tm.tm_min, at_tm.tm_sec);
+
+  return len;
+}
+
+static int
+cub_get_time_string (char *buf, struct timeval *time_val)
+{
+  struct tm tm, *tm_p;
+  time_t sec;
+  int millisec;
+
+  if (buf == NULL)
+    {
+      return 0;
+    }
+
+  if (time_val == NULL)
+    {
+      struct timeb tb;
+
+      /* current time */
+      (void) ftime (&tb);
+      sec = tb.time;
+      millisec = tb.millitm;
+    }
+  else
+    {
+      sec = time_val->tv_sec;
+      millisec = time_val->tv_usec / 1000;
+    }
+
+  tm_p = localtime_r (&sec, &tm);
+  tm.tm_mon++;
+
+  buf[0] = ((tm.tm_year % 100) / 10) + '0';
+  buf[1] = (tm.tm_year % 10) + '0';
+  buf[2] = '-';
+  buf[3] = (tm.tm_mon / 10) + '0';
+  buf[4] = (tm.tm_mon % 10) + '0';
+  buf[5] = '-';
+  buf[6] = (tm.tm_mday / 10) + '0';
+  buf[7] = (tm.tm_mday % 10) + '0';
+  buf[8] = ' ';
+  buf[9] = (tm.tm_hour / 10) + '0';
+  buf[10] = (tm.tm_hour % 10) + '0';
+  buf[11] = ':';
+  buf[12] = (tm.tm_min / 10) + '0';
+  buf[13] = (tm.tm_min % 10) + '0';
+  buf[14] = ':';
+  buf[15] = (tm.tm_sec / 10) + '0';
+  buf[16] = (tm.tm_sec % 10) + '0';
+  buf[17] = '.';
+  buf[20] = (millisec % 10) + '0';
+  millisec /= 10;
+  buf[19] = (millisec % 10) + '0';
+  millisec /= 10;
+  buf[18] = (millisec % 10) + '0';
+  buf[21] = '\0';
+
+  return 21;
+}
+
+static void
+cub_remove_char (char *string, char ch)
+{
+  for (; *string != '\0'; string++)
+    {
+      if (*string == ch)
+	{
+	  strcpy (string, string + 1);
+	  string--;
+	}
+    }
+}
+
+static int
+cub_is_ddl_type (int node)
+{
+  switch (node)
+    {
+    case PT_ALTER:
+    case PT_ALTER_INDEX:
+    case PT_ALTER_SERIAL:
+    case PT_ALTER_STORED_PROCEDURE:
+    case PT_ALTER_TRIGGER:
+    case PT_ALTER_USER:
+    case PT_CREATE_ENTITY:
+    case PT_CREATE_INDEX:
+    case PT_CREATE_SERIAL:
+    case PT_CREATE_STORED_PROCEDURE:
+    case PT_CREATE_TRIGGER:
+    case PT_CREATE_USER:
+    case PT_DROP:
+    case PT_DROP_INDEX:
+    case PT_DROP_SERIAL:
+    case PT_DROP_STORED_PROCEDURE:
+    case PT_DROP_TRIGGER:
+    case PT_DROP_USER:
+    case PT_GRANT:
+    case PT_RENAME:
+    case PT_REVOKE:
+    case PT_REMOVE_TRIGGER:
+    case PT_RENAME_TRIGGER:
+    case PT_UPDATE_STATS:
+    case PT_TRUNCATE:
+      return true;
```

Comment:
Shouldn't **PT_COMMIT**, **PT_ROLLBACK** be added to the case?
I'm not sure, what do you think?

Reply:
Since commit and rollback are performed after executing DML as well as DDL, this modification request is not desirable.
**I withdraw this comment.**